### PR TITLE
Various lua changes/cleanup

### PR DESCRIPTION
**Pull request by teto**
###### Motivation for this change
some experiment I am running:
- adding a lua lib
- change lua LUA_PATH_DEFAULT to not include /usr/local etc
- ~~see if we can install lua modules in a flat hierarchy, which implies to remove the random~~ (left for later)
```
  LUA_CPATH = "${lgi}/lib/lua/${lua.luaversion}/?.so";
  LUA_PATH  = "${lgi}/share/lua/${lua.luaversion}/?.lua;;";
```
in the code (awesome for instance) and use paths from lua.lib instead.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/top-level/lua-packages.nix [hunk 1]
```diff
@@ -132,7 +107,7 @@ with self; {
     };
   };
 
-  vicious = toLuaModule(stdenv.mkDerivation rec {
+  vicious = luaLib.toLuaModule( stdenv.mkDerivation rec {
```
Comment (SuperSandro2000) [inline pkgs/top-level/lua-packages.nix:110]: Please revert.

File: pkgs/servers/xmpp/prosody/default.nix [hunk 2]
```diff
@@ -1,19 +1,24 @@
...
 , withCommunityModules ? [ ] }:
 
-assert withLibevent -> luaevent != null;
-assert withDBI -> luadbi != null;
-
 with lib;
 
 
+let
+  luaEnv = lua.withPackages(p: with p;[
```
Comment (SuperSandro2000) [inline pkgs/servers/xmpp/prosody/default.nix:15]:
```suggestion
  luaEnv = lua.withPackages(p: with p; [
```

File: pkgs/servers/xmpp/prosody/default.nix [hunk 3]
```diff
@@ -1,19 +1,24 @@
 { stdenv, fetchurl, lib, libidn, openssl, makeWrapper, fetchhg
-, lua5, luasocket, luasec, luaexpat, luafilesystem, luabitop
+, lua
 , nixosTests
-, withLibevent ? true, luaevent ? null
+, withLibevent ? true
 , withDBI ? true, luadbi ? null
```
Comment (SuperSandro2000) [inline pkgs/servers/xmpp/prosody/default.nix:5]:
```suggestion
, withDBI ? true, luadbi
```

File: pkgs/development/lua-modules/lib.nix [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+  getDataFolder = drv:
+    "${drv.pname}-${drv.version}-rocks/${drv.pname}/${drv.version}";
+
+  /* Convert derivation to a lua module.
+    so that luaRequireModules can be run later
+  */
+  toLuaModule = drv:
+    drv.overrideAttrs( oldAttrs: {
+      # Use passthru in order to prevent rebuilds when possible.
+      passthru = (oldAttrs.passthru or {})// {
```
Comment (SuperSandro2000) [inline pkgs/development/lua-modules/lib.nix:58]:
```suggestion
      passthru = (oldAttrs.passthru or {}) // {
```

File: pkgs/development/interpreters/luajit/default.nix [hunk 5]
```diff
@@ -62,6 +62,15 @@ stdenv.mkDerivation rec {
...
     fi
+
+    {
+      echo -e '
+        #undef  LUA_PATH_DEFAULT
+        #define LUA_PATH_DEFAULT "./share/lua/${luaversion}/?.lua;./?.lua;./?/init.lua"
+        #undef  LUA_CPATH_DEFAULT
+        #define LUA_CPATH_DEFAULT "./lib/lua/${luaversion}/?.so;./?.so;./lib/lua/${luaversion}/loadall.so"
+      '
+    } >> src/luaconf.h
```
Comment (SuperSandro2000) [inline pkgs/development/interpreters/luajit/default.nix:73]:
```suggestion
    echo -e '
      #undef  LUA_PATH_DEFAULT
      #define LUA_PATH_DEFAULT "./share/lua/${luaversion}/?.lua;./?.lua;./?/init.lua"
      #undef  LUA_CPATH_DEFAULT
      #define LUA_CPATH_DEFAULT "./lib/lua/${luaversion}/?.so;./?.so;./lib/lua/${luaversion}/loadall.so"
    ' >> src/luaconf.h
```

File: pkgs/development/interpreters/lua-5/interpreter.nix [hunk 6]
```diff
@@ -31,21 +34,32 @@ self = stdenv.mkDerivation rec {
...
+  # we can't pass flags to the lua makefile because for portability, everything is hardcoded
+  postPatch = ''
+    {
+      echo -e '
+        #undef  LUA_PATH_DEFAULT
+        #define LUA_PATH_DEFAULT "./share/lua/${luaversion}/?.lua;./?.lua;./?/init.lua"
+        #undef  LUA_CPATH_DEFAULT
+        #define LUA_CPATH_DEFAULT "./lib/lua/${luaversion}/?.so;./?.so;./lib/lua/${luaversion}/loadall.so"
+      '
+    } >> src/luaconf.h
```
Comment (SuperSandro2000) [inline pkgs/development/interpreters/lua-5/interpreter.nix:55]:
```suggestion
    echo -e '
      #undef  LUA_PATH_DEFAULT
      #define LUA_PATH_DEFAULT "./share/lua/${luaversion}/?.lua;./?.lua;./?/init.lua"
      #undef  LUA_CPATH_DEFAULT
      #define LUA_CPATH_DEFAULT "./lib/lua/${luaversion}/?.so;./?.so;./lib/lua/${luaversion}/loadall.so"
    ' >> src/luaconf.h
```

File: pkgs/development/interpreters/lua-5/interpreter.nix [hunk 7]
```diff
@@ -10,7 +11,9 @@
 , staticOnly ? stdenv.hostPlatform.isStatic
 }:
 let
-luaPackages = callPackage ../../lua-modules {lua=self; overrides=packageOverrides;};
+  luaPackages = callPackage ../../lua-modules {
+    lua=self; overrides=packageOverrides;
```
Comment (SuperSandro2000) [inline pkgs/development/interpreters/lua-5/interpreter.nix:15]:
```suggestion
    lua = self;
    overrides = packageOverrides;
```

File: pkgs/applications/editors/vis/default.nix [hunk 8]
```diff
@@ -1,8 +1,11 @@
 { lib, stdenv, fetchFromGitHub, pkg-config, makeWrapper, makeDesktopItem
-, ncurses, libtermkey, lpeg, lua
+, ncurses, libtermkey, lua
 , acl ? null, libselinux ? null
 }:
 
+let
+  luaEnv = lua.withPackages(ps: [ps.lpeg]);
```
Comment (SuperSandro2000) [inline pkgs/applications/editors/vis/default.nix:7]:
```suggestion
  luaEnv = lua.withPackages(ps: [ ps.lpeg ]);
```

File: pkgs/applications/editors/vis/default.nix [hunk 9]
```diff
@@ -1,8 +1,11 @@
 { lib, stdenv, fetchFromGitHub, pkg-config, makeWrapper, makeDesktopItem
-, ncurses, libtermkey, lpeg, lua
+, ncurses, libtermkey, lua
 , acl ? null, libselinux ? null
```
Comment (SuperSandro2000) [inline pkgs/applications/editors/vis/default.nix:3]:
```suggestion
, acl, libselinux
```

File: pkgs/applications/editors/neovim/tests.nix [hunk 10]
```diff
@@ -133,14 +133,15 @@ rec {
     configure.pathogen.pluginNames = [ "vim-nix" ];
   };
 
-  nvimWithLuaPackages = wrapNeovim2 "with-lua-packages" (makeNeovimConfig {
+  nvimWithLuaPackages = wrapNeovim2 "-with-lua-packages" (makeNeovimConfig {
```
Comment (SuperSandro2000) [inline pkgs/applications/editors/neovim/tests.nix:136]:
```suggestion
  nvimWithLuaPackages = wrapNeovim2 "with-lua-packages" (makeNeovimConfig {
```

File: pkgs/top-level/lua-packages.nix (hunk 1 above)
Comment (teto) [reply pkgs/top-level/lua-packages.nix:110]: why ?

File: pkgs/applications/editors/neovim/tests.nix (hunk 10 above)
Comment (teto) [reply pkgs/applications/editors/neovim/tests.nix:136]: the original commiter asked me to do the change after I had done it myself to keep all tests consistent.

File: pkgs/applications/editors/vis/default.nix (hunk 9 above)
Comment (teto) [reply pkgs/applications/editors/vis/default.nix:3]: can't it be null if not on linux ? doesn't look related to the PR.

File: pkgs/development/interpreters/lua-5/wrapper.nix [hunk 11]
```diff
@@ -37,7 +42,12 @@ let
               rm -f "$out/bin/$prg"
               if [ -x "$prg" ]; then
                 nix_debug "Making wrapper $prg"
-                makeWrapper "$path/bin/$prg" "$out/bin/$prg" --suffix LUA_PATH ';' "$LUA_PATH"   --suffix LUA_CPATH ';' "$LUA_CPATH" ${lib.concatStringsSep " " makeWrapperArgs}
+                makeWrapper "$path/bin/$prg" "$out/bin/$prg" \
+                  --set-default LUA_PATH ";;" \
+                  --suffix LUA_PATH ';' "$NIX_LUA_PATH" \
```
Comment (teto) [inline pkgs/development/interpreters/lua-5/wrapper.nix:47]: TODO revert the NIX_ prefix

File: pkgs/top-level/lua-packages.nix (hunk 1 above)
Comment (teto) [reply pkgs/top-level/lua-packages.nix:110]: ok

File: pkgs/development/interpreters/luajit/default.nix (hunk 5 above)
Comment (teto) [reply pkgs/development/interpreters/luajit/default.nix]: as I had already tested your suggestion doesn't work, gonna revert.

**Comment by teto** [comment]
@lovek323 @rasendubi would you mind checking that doesn't break awesome ? should be good.

**Comment by teto** [comment]
I've tested manually several software. I hope it doesn't break too much stuff. Merging.

**Pull request merged**